warehouse/accounts/models.py

```diff
@@ -80,6 +81,9 @@ class User(SitemapMixin, db.Model):
         nullable=True,
     )
 
+    totp_secret = Column(Binary(length=20), nullable=True)
+    totp_provisioned = Column(Boolean, nullable=False, server_default=sql.false())
```

Comment:
NTS: I don't really like this double state here (`totp_secret` holds the user's TOTP secret key, while `totp_provisioned` is only set to true once the user verifies that they've added their device correctly by entering a code), but I'm not sure about better ways to do it:

* We could pass `totp_secret` as a cookie or hidden form parameter in `manage.account.totp-secret`, but this would allow the user/page javascript to modify the TOTP secret and potentially compromise it. 
* We could regenerate `totp_secret` each time the user refreshes `manage.account.totp-secret`, but this produces a really bad UX: the user will have to scan a new QR code each time they fail to validate their device.

Is there another place we can sequester the TOTP secret state without actually updating the user model? Alternatively, is there a way to ensure cookie integrity (i.e., that the user/page hasn't modified the cookie we sent)?

Reply:
@woodruffw you could use [this utility module](https://github.com/pypa/warehouse/blob/1fbb4ac752e68b5840b9e09b68e44a165569bfa6/warehouse/utils/crypto.py) to create a signed payload if we choose to go the cookie route.

Alternatively Redis is readily available for stashing this state.

Reply:
Summarizing a discussion @woodruffw and I had last week:

While having double state enables a user to potentially "get stuck" in the intermediate state of having a secret, but not being provisioned, any alternative where a secret key is generated and displayed, but not fully persisted could result in a bad UX scenario where a user has broken Pypi entries in their 2FA app.

For example, consider the case where:
- we generate and display a secret
- the user adds it to their device, but, for whatever reason, does not provision
- we expire that secret after some amount of time (since we do not fully persist)
- the user attempts to enable 2FA again
- we generate and display a new secret
- the user adds it to their device

The user will then have two Pypi entries in their device, of which only one is correct and usable. 

Depending on our expectations for the type of user, this might be a rare case, but in order to fully ensure good UX related to their 2FA app, we probably want to go with updating the user model when we generate the secret. Especially given that there are no real security drawbacks of having this intermediate state, it seems like a totally adequate solution to me. The only thing is that future development needs to be aware of the double totp state.